Optimization and refactor of scrollFire

yanickrochon opened the pull request:
This commit solves issue #3916 and allow `Materialize.scrollFire` to be called any number of times. The component will always use a single timer and a single scroll listener. The component will also cleanup itself if it becomes inactive, and free any unused resources.

Code review welcome!

acburst commented:
I decided to go in a different direction. I generalized the throttle function and used that instead of the setInterval.

Added in 0257ec7. Let me know if there are still problems with my approach

yanickrochon commented:
You still keep references to the options passed... This is creating memory leaks.

yanickrochon commented:
.. and you are still testing the element's bounding box before validating if you actually even need to test that element. I could've simply refactored my PR instead. You basically shut off all my PR for your own versions without discussing any matter. Do I really want to invest more time on this project for my PR?

kmmbvnr commented:
@yanickrochon Can you provide a codepen that demonstrates the issue?

yanickrochon commented:
@kmmbvnr I don't need to.

```js
Materialize.scrollFire({ selector: '.load-more', callback: function () { });
```

will always run in the background. *Always*. The component won't care if the event has been fire or not, [this line](https://github.com/Dogfalo/materialize/blob/master/js/scrollFire.js#L38) will always be executed for ever and ever. More over, during the lifecycle of the page, if another `.load-more` element is created (using React), then the component will *still* be called another time

```
Materialize.scrollFire({ selector: '.load-more', callback: function () { });
```
... it needs to, since the first one is flagged as [`done`](https://github.com/Dogfalo/materialize/blob/master/js/scrollFire.js#L29) (but is [still getting evaluated](https://github.com/Dogfalo/materialize/blob/master/js/scrollFire.js#L10) anyway!).

This PR, which is in conflict now because, instead of discussing, you guys decided to trash this one, addressed these issues.

kmmbvnr commented:
Mm, for me it looks too late to remove a scroll listener when onScroll event happened.

It's better to manage component state explicitly and remove listeners as soon as component been destroyed.

@acburst can we have an ability to destroy scrollSpy? Ex:

```javacript
  $('.selector'').scrollSpy('destroy');
```

That's like the MaterializeCSS already have for the selects `$('select').material_select('destroy');`

yanickrochon commented:
@kmmbvnr  sure! but we would need to create two constructors

1. `Materialize.scrollFire(options);` to support current implementations, keep BC, while implementing the fixes as I mentioned already. (Seriously, I want to contribute, but please don't go behind my PR and re-implement it... especially if the issues the PR addresses aren't resolved!) This API should warn that it is deprecated when used.
2. `$(selector).scrollFire(options)` the new API, bound to the element, where `$(selector).scrollFire('destroy');` may be called to remove all handlers and cleanup.
